### DISCUSSION TITLE
Compilation issues on AWS

### DISCUSSION THREAD

**Discussion by cwmac** [dtcenter/METplus#2450] [Installation] [answered]
Having an issue getting MET to compile fully on an AWS box. Getting the following errors during the ascii2nc portion of the compilation:
 
/usr/bin/ld: ascii2nc-ascii2nc.o: in function `FileHandler::setGridMask(Grid&)':
ascii2nc.cc:(.text._ZN11FileHandler11setGridMaskER4Grid[_ZN11FileHandler11setGridMaskER4Grid]+0x26): undefined reference to `MaskFilters::set_grid_mask(Grid*)'
/usr/bin/ld: ascii2nc-ascii2nc.o: in function `FileHandler::setAreaMask(TwoD_Array<bool>&)':
ascii2nc.cc:(.text._ZN11FileHandler11setAreaMaskER10TwoD_ArrayIbE[_ZN11FileHandler11setAreaMaskER10TwoD_ArrayIbE]+0x26): undefined reference to `MaskFilters::set_area_mask(TwoD_Array<bool>*)'
/usr/bin/ld: ascii2nc-ascii2nc.o: in function `FileHandler::setPolyMask(MaskPoly&)':
ascii2nc.cc:(.text._ZN11FileHandler11setPolyMaskER8MaskPoly[_ZN11FileHandler11setPolyMaskER8MaskPoly]+0x26): undefined reference to `MaskFilters::set_poly_mask(MaskPoly*)'
/usr/bin/ld: ascii2nc-ascii2nc.o: in function `FileHandler::setSIDMask(StringArray&)':
ascii2nc.cc:(.text._ZN11FileHandler10setSIDMaskER11StringArray[_ZN11FileHandler10setSIDMaskER11StringArray]+0x26): undefined reference to `MaskFilters::set_sid_mask(StringArray*)'
/usr/bin/ld: ascii2nc-file_handler.o: in function `FileHandler::FileHandler(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
file_handler.cc:(.text+0x89): undefined reference to `MaskFilters::MaskFilters()'
/usr/bin/ld: ascii2nc-file_handler.o: in function `FileHandler::_addObservations(Observation const&)':
file_handler.cc:(.text+0xf3a): undefined reference to `MaskFilters::is_filtered(double, double)'
/usr/bin/ld: file_handler.cc:(.text+0xf89): undefined reference to `MaskFilters::is_filtered_sid(char const*)'
collect2: error: ld returned 1 exit status
make[4]: *** [Makefile:489: ascii2nc] Error 1
 
Full log file is attached. Any thoughts on what could be the issue?

[make_install.log](https://github.com/dtcenter/METplus/files/13626559/make_install.log)

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hello @cwmac, I see you're running into an issue compiling MET version 11.1.0. You're getting undefined references from the linker for ascii2nc.

Thanks for sending along the `make_install.log` file. That makes it much easier to debug. Judging by its contents, I infer that you're configuring/compiling MET *without* support for Python embedding. So that means that you are not passing the `--enable-python` option to the configure script. And that's totally fine, support for Python embedding is optional and is disabled by default. However, I think this compilation error you're getting is due to the fact that `--enable-python` is not used.

We do most of our development, testing, and common installations of MET with Python embedding enabled. And that has masked this internal library dependency problem that you've uncovered. In fact, I am able to replicate the exact same linker error in ascii2nc when compiling MET 11.1.0 without `--enable-python`.

The workaround is pretty simple. Just follow [these instructions](https://met.readthedocs.io/en/latest/Users_Guide/appendixF.html#compiling-met-for-python-embedding) to compile support for Python embedding. Set the `MET_PYTHON_BIN_EXE`, `MET_PYTHON_CC`, and `MET_PYTHON_LD` environment variables and then run MET's `configure` script with the `--enable-python` option. Please let me know how that goes.

Thanks for pointing out this problem! I'll write up a bugfix issue for MET version 11.1 and make sure we include a fix in a bugfix release, as well as any future MET releases.

**Reply by JohnHalleyGotway** [reply]
FYI, here's the issue I wrote up to document this bug:
https://github.com/dtcenter/MET/issues/2760

**Comment by cwmac** [comment]
Hey @JohnHalleyGotway, was able to get MET to compile with the 'enable-python' flag. Thanks for the help!

**Reply by JohnHalleyGotway** [reply]
@cwmac thanks for confirming. I'll go ahead and mark this discussion as being answered but will work on some better checks to prevent this from happening in the future.
